src/ServiceControl.Monitoring/Infrastructure/VariableHistoryIntervalStore.cs

```diff
@@ -0,0 +1,45 @@
+﻿namespace ServiceControl.Monitoring.Infrastructure
+{
+    using System;
+    using System.Collections.Concurrent;
+    using Messaging;
+
+    public abstract class VariableHistoryIntervalStore
+    {
+        ConcurrentDictionary<HistoryPeriod, IntervalsStore> histories;
+
+        protected VariableHistoryIntervalStore()
+        {
+            histories = new ConcurrentDictionary<HistoryPeriod, IntervalsStore>();
+
+            foreach (var period in HistoryPeriod.All)
+            {
+                var intervalSize = TimeSpan.FromTicks(period.Value.Ticks / IntervalsPerStore);
+
+                histories.TryAdd(period, new IntervalsStore(intervalSize, IntervalsPerStore));
+            }
+        }
+
+        public IntervalsStore.EndpointInstanceIntervals[] GetIntervals(HistoryPeriod period, DateTime now)
+        {
+            IntervalsStore store;
+
+            if (histories.TryGetValue(period, out store))
+            {
+                return store.GetIntervals(now);
+            }
+
+            throw new Exception("Unsupported history size.");
+        }
+
+        public void Store(EndpointInstanceId instanceId, RawMessage.Entry[] entries)
+        {
+            foreach (var kvHistory in histories.ToArray())
```

Comment:
Is this `ToArray()` there to lock the collection before storing? Will it do that? We don't actually need to as we never mess with the `histories` collection after we create it. Can we just use `histories.Values` instead?